tokenclient/tokenFlows.go

```diff
@@ -0,0 +1,177 @@
+	}
+	for name, value := range options.Params {
+		data.Set(name, value) // potentially overwrites data which was set before
+	}
+	targetURL, err := t.getURL(customerTenantURL)
+	if err != nil {
+		return "", err
+	}
+	r, e := http.NewRequestWithContext(ctx, http.MethodPost, targetURL, strings.NewReader(data.Encode())) // URL-encoded payload
+	if e != nil {
+		return "", fmt.Errorf("error performing client credentials flow: %w", e)
+	}
+	r.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	tokenJSON, err := t.performRequest(r)
+	if err != nil {
+		return "", err
+	}
+	var response tokenResponse
+	_ = json.Unmarshal(tokenJSON, &response)
+	if response.Token == "" {
+		return "", fmt.Errorf("error parsing requested client credential token: %v", string(tokenJSON))
+	}
+	return response.Token, nil
+}
+
+func (t *TokenFlows) getURL(customerTenantHost string) (string, error) {
+	if customerTenantHost == "" {
+		return t.tokenURI, nil
+	}
+	customHost, err := url.Parse(customerTenantHost)
+	if err == nil && customHost.Host != "" {
+		return "https://" + customHost.Host + tokenEndpoint, nil
+	}
+	return "", fmt.Errorf("customer tenant host '%v' can't be accepted: %v", customerTenantHost, err)
+}
+
+func (t *TokenFlows) performRequest(r *http.Request) ([]byte, error) {
+	res, err := t.options.HTTPClient.Do(r)
+	if err != nil {
+		return nil, fmt.Errorf("request to '%v' failed: %w", r.URL, err)
+	}
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("request to '%v' failed with status code '%v' and payload: '%v'", r.URL, res.StatusCode, string(body))
+	}
+	if err != nil || body == nil || !json.Valid(body) {
+		return nil, fmt.Errorf("request to '%v ' provides no valid json content: %w", r.URL, err)
+	}
+	return body, nil
+}
+
+func defaultTLSConfig(identity env.Identity) (*tls.Config, error) {
+	certPEMBlock := []byte(identity.GetCertificate())
+	keyPEMBlock := []byte(identity.GetKey())
+
+	tlsCert, err := tls.X509KeyPair(certPEMBlock, keyPEMBlock)
+	if err != nil {
+		return nil, fmt.Errorf("error creating x509 key pair for defaultTLSConfig: %w", err)
+	}
+	tlsCertPool, err := x509.SystemCertPool()
+	if err != nil {
+		return nil, fmt.Errorf("error setting up cert pool for defaultTLSConfig: %w", err)
+	}
+	ok := tlsCertPool.AppendCertsFromPEM(certPEMBlock)
+	if !ok {
+		return nil, fmt.Errorf("error adding certs to pool for defaultTLSConfig: %w", err)
+	}
+	tlsConfig := &tls.Config{
+		MinVersion:   tls.VersionTLS12,
+		RootCAs:      tlsCertPool,
+		Certificates: []tls.Certificate{tlsCert},
+	}
+	return tlsConfig, nil
+}
+
+// TODO avoid duplication
+func defaultHTTPClient(tlsConfig *tls.Config) *http.Client {
+	client := &http.Client{
+		Timeout: time.Second * 10, // TODO check
+	}
+	if tlsConfig != nil {
+		client.Transport = &http.Transport{
+			TLSClientConfig:     tlsConfig,
+			MaxIdleConnsPerHost: 50, // TODO check
```

Comment:
Since the customer tenantid is part of the host I think this value is too high. Instead or additionally maybe limit the `MaxIdleConns`?